manifests/init.pp

```diff
@@ -34,21 +34,39 @@
     Hash $pkg_list,
 ) {
 
-    # Default value (for bare package names)
-    $default = {'ensure' => 'installed'}
+  $script_name="puppet-profile_additional_packages-packages_without_params.sh"
 
-    # Limit list of packages by OS Family
-    $packages = $pkg_list[ $facts['os']['family'] ]
+  file { "/root/scripts/${script_name}":
+    ensure  => 'present',
+    mode    => '0740',
+    owner   => 'root',
+    group   => 'root',
+    source  => "puppet:///modules/${module_name}/root/scripts/${script_name}",
+    require => File['/root/scripts'],
+  }
 
-    if $packages =~ Hash[ String[1], Data, 1 ] {
-      # Find keys without a value and add default value
-      $sane_pkg_list = $packages.map |$key, $val| {
-          if $val { [ $key, $val ] }
-          else { [ $key, $default ] }
-      }.convert_to(Hash)
+  # Limit list of packages by OS Family
+  $packages = $pkg_list[ $facts['os']['family'] ]
 
-      ensure_packages( $sane_pkg_list )
+  if $packages =~ Hash[ String[1], Data, 1 ] {
+
+    # Find keys without a value and install those pacakges
+    # via exec/'yum install' (single yum transaction)
+    $packages_without_params = $packages.filter |$key, $val| {
+      empty($val)
+    }.keys.join(' ')
+    exec { 'install_packages_without_params':
+      command => "/root/scripts/${script_name} ${packages_without_params}",
+      unless  => "/usr/bin/rpm -q ${packages_without_params}" 
     }
 
-}
+    # Install keys that DO have values using ensure_packages
```

Comment:
I had presumed we would still run `ensure_packages($packages)` across the whole Hash just to catch anything that may have slipped through the cracks.

I presumed it would slow much of anything down if the package is already installed. But I could be wrong.

Did you test that?

Reply:
I didn't test that. I think in the exec, if dnf failed to install 1+ packages, it'd cancel or even possibly revert the transaction. For example, if you specify a package name that is not available in any repos or if it cannot download a package I know it bails out. So I _think_ that it'd either succeed 100% or fail.

So we could do what you're suggesting and:
- if the exec succeeds, then I agree that the ensure_packages shouldn't hardly take longer since most everything will already be installed;
- if the exec fails, it'd be an effective safeguard and should install everything, but it'd take longer as it'll proceed one-by-one like it currently does, and at the same time we might not notice that there's an issue

Or we could leave it as-is and it might be more obvious it's a problem?

I'm OK to go either way. It's easy to do what you're saying. Thoughts?

Reply:
The script should return an error if failed and that would be logged in puppet output as an error by the exec.